psyneulink/library/compositions/pytorchcomponents.py

```diff
@@ -0,0 +1,257 @@
+
+            f_x = builder.call(exp, [arg])
+            f_x = builder.fadd(one, f_x)
+            f_x = builder.fdiv(one, f_x)
+
+            ret = builder.fsub(one, f_x)
+            ret = builder.fmul(f_x, ret)
+            return ret
+
+    else:
+        raise Exception(
+            f"Function type {node.function} is currently unsupported by compiled execution!")
+
+    # do computations
+    with pnlvm.helpers.for_loop_zero_inc(builder, dim, "derivative_loop") as (builder, iterator):
+        val_ptr = builder.gep(input_vector, [iterator])
+        val = builder.load(val_ptr)
+        val = modify_value(val)
+        output_location = builder.gep(output_vector, [iterator])
+        builder.store(val, output_location)
+
+    builder.ret_void()
+
+    return llvm_func
+
+
+class PytorchMechanism():
+    """
+    An interpretation of a mechanism as an equivalent pytorch object
+    """
+    def __init__(self, mechanism, component_idx, device, context=None):
+        self._mechanism = mechanism
+        self._idx = component_idx
+        self._context = context
+
+        self.function = pytorch_function_creator(mechanism.function, device, context)
+        self._context = context
+        self.value = None
+        self._llvm_value = None
+        self.afferents = []
+        self.efferents = []
+
+        self._target_mechanism = None
+
+    def add_efferent(self, efferent):
+        if not efferent in self.efferents:
+            self.efferents.append(efferent)
+
+    def add_afferent(self, afferent):
+        if not afferent in self.afferents:
+            self.afferents.append(afferent)
+
+
+    def collate_afferents(self):
+        """
+        Returns weight-multiplied sum of all afferent projections
+        """
+        return sum([proj.execute(proj.sender.value) for proj in self.afferents])
+
+    def execute(self, variable):
+        self.value = self.function(variable)
+        # print(f"{self} input: {variable}\n")
+        # print(f"{self} output: {self.value}\n")
+        return self.value
+
+    def execute_llvm(self, ctx, builder, state, params, mech_input, data):
+        mech_func = ctx.import_llvm_function(self._mechanism)
+
+        mech_param = builder.gep(params, [ctx.int32_ty(0),
+                                                  ctx.int32_ty(0),
+                                                  ctx.int32_ty(self._idx)])
+
+        mech_state = builder.gep(state, [ctx.int32_ty(0),
+                                            ctx.int32_ty(0),
+                                            ctx.int32_ty(self._idx)])
+
+        mech_output = builder.gep(data, [ctx.int32_ty(0),
+                                            ctx.int32_ty(0),
+                                            ctx.int32_ty(self._idx)])
+
+        builder.call(mech_func, [mech_param, mech_state,
+                                    mech_input, mech_output])
+        
+        for i in range(len(mech_input.type.pointee.elements[0])):
```

Comment:
`enumerate(x)` isn't possible in this case, since the type is of ir.types.ArrayType, which is not iterable. `len()` in this case actually just uses a magic method on the class, which is why it works.

I'll make it more explicit by moving the `len()` call out of the `range()` call.

Reply:
ir.ArrayType has `count` property, that should be more readable than `range(len(` (even if split)